internal/handler/workloadidentity/handler.go

```diff
@@ -0,0 +1,124 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package workloadidentity
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/go-logr/logr"
+
+	"github.com/gardener/gardener-discovery-server/internal/utils"
+)
+
+const (
+	headerCacheControl = "Cache-Control"
+	pubCacheControl    = "public, max-age=3600"
+
+	headerContentType = "Content-Type"
+	mimeAppJSON       = "application/json"
+
+	responseMethodNotAllowed = `{"code":405,"message":"method not allowed"}`
+)
+
+// Handler implements handler functions for the openid configuration and JWKS endpoints.
+type Handler struct {
+	oidc []byte
+	jwks []byte
+	log  logr.Logger
+}
+
+// New creates new workload identity handler.
+func New(openIDConfig, jwks []byte, logger logr.Logger) (*Handler, error) {
+	conf, err := utils.LoadOpenIDConfig(openIDConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load openid configuration: %w", err)
+	}
+
+	issuerURL, err := url.Parse(conf.Issuer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse issuer url, %w", err)
+	}
+	if issuerURL.Scheme != "https" {
+		return nil, errors.New("invalid issuer url scheme")
+	}
+	if issuerURL.RawQuery != "" {
+		return nil, errors.New("issuer url must not contain query")
+	}
+	if issuerURL.Fragment != "" {
+		return nil, errors.New("issuer url must not contain fragment")
+	}
+
+	jwksURL, err := url.Parse(conf.JWKSURI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse jwks url, %w", err)
+	}
+	if jwksURL.Scheme != "https" {
+		return nil, errors.New("invalid jwks url scheme")
+	}
+
+	keySet, err := utils.LoadKeySet(jwks)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load json web key set: %w", err)
+	}
+
+	for _, k := range keySet.Keys {
+		if !k.IsPublic() {
+			return nil, fmt.Errorf("jwks key with id %q is not public", k.KeyID)
+		}
+	}
+
+	return &Handler{
+		oidc: openIDConfig,
+		jwks: jwks,
+		log:  logger,
+	}, nil
+}
+
+// HandleWellKnown handles /.well-known/openid-configuration.
+func (h *Handler) HandleWellKnown(w http.ResponseWriter, r *http.Request) {
+	handleRequest(h.log.WithName("well-known"), h.oidc, w, r)
+}
+
+// HandleJWKS handles JWKS response.
+func (h *Handler) HandleJWKS(w http.ResponseWriter, r *http.Request) {
+	handleRequest(h.log.WithName("jwks"), h.jwks, w, r)
+}
+
+func handleRequest(logger logr.Logger, responseData []byte, w http.ResponseWriter, r *http.Request) {
+	if w.Header().Get("Strict-Transport-Security") == "" {
+		w.Header().Set("Strict-Transport-Security", "max-age=31536000")
+	}
+
+	if isValid := handleInvalidRequest(logger.WithName("invalid-request-handler"), w, r); !isValid {
```

Comment:
I think that this named handler is too concrete and is implementation detail related. I would rather enhance the error log so we know where the error originates from.

The common functions of both `workloadidentity` and `openidmeta` handlers can be extracted as middlewares in the `handlers` package. It is not necessary that this is done with this PR. We can do it in a separate PR if you want.

```go
const (
	headerContentType = "Content-Type"
	mimeAppJSON       = "application/json"
)

func SetHSTS(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if w.Header().Get("Strict-Transport-Security") == "" {
			w.Header().Set("Strict-Transport-Security", "max-age=31536000")
		}
		h.ServeHTTP(w, r)
	})
}

func AllowMethods(h http.Handler, log logr.Logger, allowedMethods ...string) http.Handler {
	var (
		responseMethodNotAllowed = []byte(`{"code":405,"message":"method not allowed"}`)
		methods                  = map[string]struct{}{}
	)
	for _, m := range allowedMethods {
		methods[m] = struct{}{}
	}
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, ok := methods[r.Method]; !ok {
			w.Header().Set(headerContentType, mimeAppJSON)
			w.WriteHeader(http.StatusMethodNotAllowed)
			if _, err := w.Write([]byte(responseMethodNotAllowed)); err != nil {
				log.Error(err, "Failed writing response")
				return
			}
			return
		}
		h.ServeHTTP(w, r)
	})
}

func HandleNotFound(log logr.Logger) http.Handler {
	var responseNotFound = []byte(`{"code":404,"message":"not found"}`)
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set(headerContentType, mimeAppJSON)
		w.WriteHeader(http.StatusNotFound)
		if _, err := w.Write([]byte(responseNotFound)); err != nil {
			log.Error(err, "Failed writing response")
			return
		}
	})
}

```

Reply:
With 3ce6688de37abb6be02e774dc1a37eb76df001ff workloadidentity and openidmeta handlers are reusing as much as possible middleware handlers.